source/ice_api.c

```diff
@@ -0,0 +1,1835 @@
+            pIceCandidate->priority = Ice_ComputeCandidatePriority( pIceCandidate );
+        }
+
+        if( retStatus == ICE_RESULT_OK )
+        {
+            *ppCandidate = pIceCandidate;
+        }
+    }
+
+    return retStatus;
+}
+/*------------------------------------------------------------------------------------------------------------------*/
+
+/* Ice_AddSrflxCandidate - The application calls this API for adding Server Reflexive candidate. */
+
+IceResult_t Ice_AddSrflxCandidate( IceAgent_t * pIceAgent,
+                                   const IceIPAddress_t ipAddr,
+                                   IceCandidate_t ** ppCandidate,
+                                   uint8_t * pTransactionIdBuffer,
+                                   uint8_t ** ppSendStunMessageBuffer,
+                                   uint32_t * pSendStunMessageBufferLength )
+{
+    IceResult_t retStatus = ICE_RESULT_OK;
+    IceCandidate_t * pIceCandidate;
+    int localCandidateCount;
+
+    if( ( pIceAgent == NULL ) ||
+        ( ppCandidate == NULL ) ||
+        ( ppSendStunMessageBuffer == NULL ) ||
+        ( pSendStunMessageBufferLength == NULL ) ||
+        ( pTransactionIdBuffer == NULL ) )
+    {
+        retStatus = ICE_RESULT_BAD_PARAM;
+    }
+    else
+    {
+        localCandidateCount = Ice_GetValidLocalCandidateCount( pIceAgent );
+
+        if( localCandidateCount >= ICE_MAX_LOCAL_CANDIDATE_COUNT )
+        {
+            retStatus = ICE_RESULT_MAX_CANDIDATE_THRESHOLD;
+        }
+        else
+        {
+            pIceCandidate = &( pIceAgent->localCandidates[ localCandidateCount ] );
+        }
+
+        if( retStatus == ICE_RESULT_OK )
+        {
+            pIceCandidate->isRemote = 0;
+            memcpy( &( pIceCandidate->ipAddress ),
+                    &( ipAddr ),
+                    sizeof( IceIPAddress_t ) );
+            pIceCandidate->iceCandidateType = ICE_CANDIDATE_TYPE_SERVER_REFLEXIVE;
+            pIceCandidate->state = ICE_CANDIDATE_STATE_NEW;
+            pIceCandidate->priority = Ice_ComputeCandidatePriority( pIceCandidate );
+
+            retStatus = Ice_CreateRequestForSrflxCandidate( pIceAgent,
+                                                            &( pIceAgent->stunMessageBuffer[ 0 ] ),
+                                                            pTransactionIdBuffer,
+                                                            pSendStunMessageBufferLength );
+
+            if( retStatus == ICE_RESULT_OK )
+            {
+                *ppSendStunMessageBuffer = &( pIceAgent->stunMessageBuffer[ 0 ] );
+                *ppCandidate = pIceCandidate;
+            }
+        }
+    }
+
+    return retStatus;
+}
+/*------------------------------------------------------------------------------------------------------------------*/
+
+/* Ice_AddRemoteCandidate - The application calls this API for adding remote candidates. */
+
+IceResult_t Ice_AddRemoteCandidate( IceAgent_t * pIceAgent,
```

Comment:
Is this a function called internally, if no, then NULL check will have to be added for all arguments.